dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+                        let cont = env[env.len() - 1 - *debruijn_index].cont;
+                        cfg.redirect(node, cont);
+                    }
+                }
+            }
+        }
+    }
+
+    fn to_target_inner(&self, parent_cont: Target, node: Index) -> Target {
+        let node = &self[node];
+        let cont = match node.next {
+            Some(i) => self.to_target_inner(parent_cont, i),
+            None => parent_cont,
+        };
+
+        match &node.expr {
+            Ir::Done => Target::Done,
+            Ir::Recv(t) => Target::Recv(t.clone(), Rc::new(cont)),
+            Ir::Send(t) => Target::Send(t.clone(), Rc::new(cont)),
+            Ir::Call(callee) => {
+                let callee = self.to_target_inner(Target::Done, *callee);
+                Target::Call(Rc::new(callee), Rc::new(cont))
+            }
+            Ir::Split(tx_only, rx_only) => {
+                let (tx_only, rx_only) = (*tx_only, *rx_only);
+                let tx_target = self.to_target_inner(cont.clone(), tx_only);
+                let rx_target = self.to_target_inner(cont, rx_only);
+                Target::Split(Rc::new(tx_target), Rc::new(rx_target))
+            }
+            Ir::Choose(choices) => {
+                let targets = choices
+                    .iter()
+                    .map(|&choice| self.to_target_inner(cont.clone(), choice))
+                    .collect();
+                Target::Choose(targets)
+            }
+            Ir::Offer(choices) => {
+                let targets = choices
+                    .iter()
+                    .map(|&choice| self.to_target_inner(cont.clone(), choice))
+                    .collect();
+                Target::Offer(targets)
+            }
+            Ir::Loop(_, body) => Target::Loop(Rc::new(self.to_target_inner(cont, *body))),
+            Ir::IndexedBreak(_) | Ir::LabeledBreak(_) => {
+                panic!("uneliminated break in CFG")
+            }
+            Ir::LabeledContinue(_) => panic!("uneliminated labeled continue in CFG"),
+            Ir::IndexedContinue(i) => {
+                debug_assert!(node.next.is_none(), "continue must be the end of a block");
+                Target::Continue(*i)
+            }
+            Ir::Type(t) => {
+                // Optimize a little bit by doing the `Then` transform ourselves if the next
+                // continuation is a `Done`.
+                match cont {
+                    Target::Done => Target::Type(t.clone()),
+                    _ => Target::Then(Rc::new(Target::Type(t.clone())), Rc::new(cont)),
+                }
+            }
+            Ir::Error => match cont {
+                Target::Done => Target::Error,
+                _ => Target::Then(Rc::new(Target::Error), Rc::new(cont)),
+            },
+        }
+    }
+
+    fn to_target(&self, node: Index) -> Target {
+        self.to_target_inner(Target::Done, node)
+    }
+
+    fn drain_errors(&mut self) -> Option<syn::Error> {
```

Comment:
This is not idempotent and it does not consume `self`, and that worries me a little, because it means that it would be possible to drain all the errors, leaving behind an (apparently) error-less `Cfg` which nevertheless contains `Ir::Error` nodes. I would recommend changing the API to prevent this. As noted elsewhere, I think an elegant solution might be to propagate full error information in `Ir::Error` and `Target::Error` nodes, rather than tracking it separately, and only extract it at the end.

Reply:
Yes, same concerns here. There are some issues w/ errors on nodes which may have child nodes w/ errors where we can't turn them into error nodes themselves though. I'll think about how to fix that today.